### PR TITLE
fix: upsert should find unique index created by one-to-one relation

### LINKED ISSUE #8587: Upsert expects unique constraint for `OneToOne` columns which imply uniqueness already
I get this error:
```
An upsert requires conditions that have a unique constraint but none was found for conflict properties: original_response_id
```
For code like this:
```typescript
import TypeORM from 'typeorm';
import { Response } from './response';

@TypeORM.Entity()
export class Archive {
	@TypeORM.PrimaryGeneratedColumn('uuid')
		id!: string;

	@TypeORM.OneToOne(() => Response, response => response.id)
	@TypeORM.JoinColumn()
		originalResponse!: Response;
}
```

```typescript
		await this._archiveRepository.upsert(archive, [ 'original_response_id' ]);
```

I also tried changing `'original_response_id'` to `'original_response'` or `'originalResponse'` and it did not help.

This is the line where error is thrown:
https://github.com/typeorm/typeorm/blob/26581d0264268045496b43f7a0f338eb9f2b70cc/src/entity-manager/EntityManager.ts#L509-L511

### PR DESCRIPTION
**Pull request by joeflateau**
### Description of change

Fixes #8587 

One-to-one relations create a unique index, that index should be suitable for an upsert conflict target.

### Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [x] `npm run lint` passes with this change
- [x] `npm run test` passes with this change
- [x] This pull request links relevant issues as `Fixes #0000`
- [x] There are new or updated unit tests validating the change
- [ ] Documentation has been updated to reflect this change
- [x] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: src/entity-manager/EntityManager.ts [hunk 1]
```diff
@@ -495,10 +496,16 @@ export class EntityManager {
...
         }
 
+        const relationMetadatas = metadata.relations.map(relation => (relation.embeddedMetadata ?? relation.entityMetadata));
+
         const uniqueColumnConstraints = [
             metadata.primaryColumns,
             ...metadata.indices.filter(ix => ix.isUnique).map(ix => ix.columns),
-            ...metadata.uniques.map(uq => uq.columns)
+            ...metadata.uniques.map(uq => uq.columns),
+            ...OrmUtils.flatten([
```
Comment (Ginden) [inline src/entity-manager/EntityManager.ts:505]: No need to `.flatten`, just spread them directly.

**Comment by pleerock** [comment]
Thanks for the fix! Quick question: do we really need a runtime check for column uniqueness? Usually we delegate everything the db, and if there is something that db cannot handle - it throws an error.

**Comment by joeflateau** [comment]
@pleerock I think it's important. It's really there to protect MySQL/MariaDB users since those DB engines don't let you specify the conflict columns. If you "upsert" and there's no unique constraint for the columns you specify as your conflict columns then Postgres will throw an error but MySQL will just insert it and move on.

**Comment by pleerock** [comment]
if that's natural MySQL behaviour, and users use MySQL's UPSERT with it's natural behaviour (orm is just a wrapper around what rdbms provides) then why do we need to prevent users from doing what they wanna do?

**Comment by joeflateau** [comment]
my philosophy is that I like to catch footguns like this where the abstraction of "upsert" could cause some unexpected behavior due to db engine differences. your call though. I can remove the check altogether if you prefer that

**Comment by pleerock** [comment]
The problem is that we definitely cannot catch all the errors - and when we handle some errors one way and other errors another way - it brings a confusion.

Also, there are users who use ORM with exist database without entities synchronization. Such users don't define entity  indices and uniques. They simply define entities and execute queries against exist database. Current approach fail for them.

> I can remove the check altogether if you prefer that

Yes, I would prefer these checks to be handles by database if it's possible. If you don't have a time or wish to do that - no worries, I'll merge this PR for now and someday we might refactor this code.

**Comment by pleerock** [comment]
thank you a lot

**Comment by Robert-Rendell-THE** [comment]
For anyone that wants typing, I did this:
```typescript
const conflictPaths: (keyof MyEntity)[] = ['columnA', 'columnB']
const opts: UpsertOptions<MyEntity> = { conflictPaths };
myEntityRepo.upsert({ ... }, opts)

**Pull request merged**